Intersection and Union support for sets

xfbs opened the pull request:
This PR adds the ability to determine both the intersection as well as the union of two sets. It attempts to be relatively generic, working for any iterator pairs that produce sorted, but not necessarily non-overlapping ranges. 

It tries to be efficient by doing the least amount of work necessary, and correct in that it will join overlapping ranges correctly. 

Correctness is verified using randomized test cases with proptest. So far, there are test cases for:

- [x] correctness of union: the union of two sets contains a value if and only if either or both of the two sets contain it
- [x] correct merging of ranges: the union iterator of sets never produces overlapping ranges
- [x] correctness of intersection: the intersection of two sets contains a value if and only if both of the two sets contain it
- [x] correct merging of intersection: the intersection iterator never produces overlapping ranges

By validating these four assumptions, we have a guarantee that the code is correct.

xfbs commented:
Edit: this PR is now ready for review. I've added a module called `operations` for these iterators. I have achieved 100% test coverage using `proptest`. I chose to make the iterators and their `new` functions public, so that you could instantiate them with another iterator. The `RangeSet::union` and `RangeSet::intersection` functions are just helper functions that do the right thing.

<img width="797" alt="Screenshot 2024-01-31 at 18 56 29" src="https://github.com/jeffparsons/rangemap/assets/786420/053a605f-77ce-4b18-9eba-6b49b67d2cfa">

xfbs commented:
Edit: yesterday I pushed another commit which implements `BitOr` and `BitAnd` for the set types using these union and intersection functionality. I took inspiration from `BTreeSet` which implements these, such that you can write:

```rust
use rangemap::RangeSet;

let left = RangeSet::from([0..100, 200..300]);
let right = RangeSet::from([100..200, 300..400]);

// get intersection as new rangeset
let union = left & right;
assert_eq!(union, BTreeSet::from([0..400]));
```

I think this is quite neat, actually I did not know that `BTreeSet` could do this before.

jeffparsons commented on src/lib.rs:
```diff
@@ -134,6 +134,7 @@ extern crate alloc;
 pub mod inclusive_map;
 pub mod inclusive_set;
 pub mod map;
+pub mod operations;
```
Is this only public because it contains `Union` and `Intersection` iterators? I think I'd prefer to keep `operations` private, and wrap those very generic iterator structs up in a simpler, more purpose-specific struct for the public API.

I.e. instead of this in the public API:

```
pub struct Union<'a, T, L, R = L> { // ...
```

Have separate...

```
pub struct Union<T> { // ...
```

for each kind of set.

(Then we can separately think about whether the more flexible version is useful to have in the public API.)

xfbs replied:
Makes sense — I had them public at first because they are generic over the iterator type and could thus be used with other iterators that emit ranges. However, I agree that it makes sense to keep things simple first and only export the types we are actually using.

jeffparsons commented on src/inclusive_set.rs:
```diff
@@ -192,6 +193,19 @@ where
+        Union::new(self.iter(), other.iter())
+    }
+
+    /// Iterator over the intersection of two range sets.
```
```suggestion
    /// Return an iterator for the intersection of this set and `other`.
```

jeffparsons commented on src/inclusive_set.rs:
```diff
@@ -192,6 +193,19 @@ where
         self.overlapping(range).next().is_some()
     }
+
+    /// Iterator over the union of two range sets.
```
```suggestion
    /// Return an iterator for the union of this set and `other`.
```

jeffparsons commented on src/set.rs:
```diff
@@ -78,6 +79,16 @@ where
         self.rm.is_empty()
     }
 
+    /// Iterator over the intersection of two range sets.
```
```suggestion
    /// Return an iterator for the intersection of this set and `other`.
```

jeffparsons commented on src/set.rs:
```diff
@@ -78,6 +79,16 @@ where
+        Intersection::new(self.iter(), other.iter())
+    }
+
+    /// Iterator over the union of two range sets.
```
```suggestion
    /// Return an iterator for the union of this set and `other`.
```

jeffparsons commented on src/operations.rs:
```diff
@@ -0,0 +1,246 @@
+
+/// Range which can be merged with a next range if they overlap.
+trait RangeIntersect: Sized {
+    /// Attempt to merge the next range into the current range, if they overlap.
```
```suggestion
    /// Return the intersection of this range and `other`, if they overlap.
```

jeffparsons reviewed (changes requested):
Mostly tiny doc comment suggestions, but one bigger picture thing about what iterator interfaces to expose in the public API.

Everything else looks really great to me — thanks for this and sorry for the slow review! 💖 🎉

xfbs commented:
Hey Jeff, no worries about the review timing :) I'm happy I can contribute a bit to this awesome crate.

I've implemented all of your suggestions and I've rebased it to fix the merge conflics.
